pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/RealtimeSegmentDataManager.java

```diff
@@ -703,9 +703,20 @@ public void run() {
         //   persisted.
         // Take upsert snapshot before starting consuming events
         if (_partitionUpsertMetadataManager != null) {
-          _partitionUpsertMetadataManager.takeSnapshot();
-          // If upsertTTL is enabled, we will remove expired primary keys from upsertMetadata after taking snapshot.
-          _partitionUpsertMetadataManager.removeExpiredPrimaryKeys();
+          if (_tableConfig.getUpsertMetadataTTL() > 0) {
+            // If upsertMetadataTTL is enabled, we will remove expired primary keys from upsertMetadata
+            // AFTER taking a snapshot. Taking the snapshot first is crucial to ensure we capture the final
+            // state of a particular key before it exits the TTL window.
```

Comment:
curious why it is critical to take snapshot first? The states of those particular keys will be gone in the next snapshot, and would that cause issue?

Reply:
> curious why it is critical to take snapshot first? The states of those particular keys will be gone in the next snapshot, and would that cause issue?

I don't see this as a very strict requirement, but let's consider the scenario where we reverse the order and the _metadataTTL logic removes keys older than the TTL window. If a server restart occurs before snapshotting, the addSegment flow will not find a snapshot for an out-of-TTL segment and will consider all records as valid for querying. However, some records may have had updates and should not be served from this out-of-TTL segment. The snapshot logic helps prevent this issue. This is my understanding from the code, but @deemoliu can confirm.

Reply:
> I don't see this as a very strict requirement

I mention this because snapshotting isn't always reliable. For instance, we've encountered Path not found exceptions during the snapshotting of some segments in the past. However, _metadataTTL would still proceed regardless. So IIUC the ordering might be more of an optimization to ensure better data correctness.

Reply:
I see, that makes sense. Based on the metadata TTL related code, it seems like taking snapshot after removing metadata out of the TTL wouldn’t affect data/query correctness but incurred some extra overhead if server restarted before taking new snapshot as it had to add metadata (already out of TTL) back to Map, however those metadata would be removed again. 

IIUC, maybe update the comment a bit that this is mainly to avoid such overhead upon unexpected server restart, but not affecting correctness.

Reply:
> it seems like taking snapshot after removing metadata out of the TTL wouldn’t affect data/query correctness but incurred some extra overhead if server restarted before taking new snapshot as it had to add metadata (already out of TTL) back to Map, however those metadata would be removed again.

Umm actually no, if a segment is out-of-TTL window during server restart then we don't follow the `doAddSegment` flow for them. So it should be a data correctness issue afaiu.

```
if (maxComparisonValue.doubleValue() < _largestSeenComparisonValue.get() - _metadataTTL) {
        _logger.info("Skip adding segment: {} because it's out of TTL", segmentName);
        MutableRoaringBitmap validDocIdsSnapshot = immutableSegment.loadValidDocIdsFromSnapshot();
        if (validDocIdsSnapshot != null) {
          MutableRoaringBitmap queryableDocIds = getQueryableDocIds(segment, validDocIdsSnapshot);
          immutableSegment.enableUpsert(this, new ThreadSafeMutableRoaringBitmap(validDocIdsSnapshot),
              queryableDocIds != null ? new ThreadSafeMutableRoaringBitmap(queryableDocIds) : null);
        } else {
          _logger.warn("Failed to find snapshot from segment: {} which is out of TTL, treating all documents as valid",
              segmentName);
        }
        return;
}
```

Reply:
Updated javadocs by adding more details.